File: Robot Control Course/Chapter8/Lesson1/Chapter8_Lesson1.py

import numpy as np

# True parameters (unknown to the controller)
m_true = 2.0   # inertia
b_true = 0.4   # viscous friction
g0_true = 5.0  # gravity coefficient

# Controller parameter estimates (to be adapted)
m_hat = 1.0
b_hat = 0.2
g0_hat = 3.0

k_p = 20.0
k_d = 8.0

def desired_trajectory(t):
    q_d = 0.5 * np.sin(0.5 * t)
    dq_d = 0.25 * np.cos(0.5 * t)
    ddq_d = -0.125 * np.sin(0.5 * t)
    return q_d, dq_d, ddq_d

def plant_dynamics(t, state, tau):
    q, dq = state
    ddq = (tau - b_true * dq - g0_true * np.sin(q)) / m_true
    return np.array([dq, ddq])

def control_law(t, q, dq, m_hat, b_hat, g0_hat):
    q_d, dq_d, ddq_d = desired_trajectory(t)
    e = q - q_d
    de = dq - dq_d

    # reference acceleration (as in computed torque)
    ddq_r = ddq_d - k_d * de - k_p * e
    dq_r = dq_d - k_p * e  # simple choice for illustration

    tau = m_hat * ddq_r + b_hat * dq_r + g0_hat * np.sin(q)
    return tau, e, de

def update_parameters(m_hat, b_hat, g0_hat, Y, tau, tau_hat, gamma=0.1):
    """
    Simple gradient update (for illustration only; rigorous laws in later lessons).
    tau: actual torque (what was commanded)
    tau_hat: Y theta_hat (model-predicted torque)
    """
    error = tau - tau_hat  # scalar
    grad = -Y * error      # gradient wrt theta_hat
    m_hat_new = m_hat - gamma * grad[0]
    b_hat_new = b_hat - gamma * grad[1]
    g0_hat_new = g0_hat - gamma * grad[2]
    return m_hat_new, b_hat_new, g0_hat_new

dt = 0.001
T = 5.0
steps = int(T / dt)

q = 0.0
dq = 0.0
history = []

for k in range(steps):
    t = k * dt
    tau, e, de = control_law(t, q, dq, m_hat, b_hat, g0_hat)

    # Regressor for this simple system
    q_d, dq_d, ddq_d = desired_trajectory(t)
    ddq_r = ddq_d - k_d * de - k_p * e
    dq_r = dq_d - k_p * e
    Y = np.array([ddq_r, dq_r, np.sin(q)])

    tau_hat = Y @ np.array([m_hat, b_hat, g0_hat])

    # Update parameter estimates (toy adaptation)
    m_hat, b_hat, g0_hat = update_parameters(m_hat, b_hat, g0_hat, Y, tau, tau_hat)

    # Integrate plant
    dq_mid = dq + 0.5 * dt * plant_dynamics(t, (q, dq), tau)[1]
    ddq_mid = (tau - b_true * dq_mid - g0_true * np.sin(q)) / m_true
    q += dq * dt
    dq += ddq_mid * dt

    history.append((t, q, dq, e, m_hat, b_hat, g0_hat))

File: Robot Control Course/Chapter8/Lesson1/Chapter8_Lesson1.java

public class AdaptiveJoint1DOF {
    static class Params {
        double mTrue = 2.0;
        double bTrue = 0.4;
        double g0True = 5.0;
        double mHat = 1.0;
        double bHat = 0.2;
        double g0Hat = 3.0;
    }

    static double[] desiredTrajectory(double t) {
        double qd = 0.5 * Math.sin(0.5 * t);
        double dqd = 0.25 * Math.cos(0.5 * t);
        double ddqd = -0.125 * Math.sin(0.5 * t);
        return new double[]{qd, dqd, ddqd};
    }

    static double controlLaw(double t, double q, double dq,
                             Params p, double kP, double kD,
                             double[] Y, double[] tauHatOut) {
        double[] traj = desiredTrajectory(t);
        double qd = traj[0], dqd = traj[1], ddqd = traj[2];
        double e = q - qd;
        double de = dq - dqd;
        double ddq_r = ddqd - kD * de - kP * e;
        double dq_r = dqd - kP * e;

        Y[0] = ddq_r;
        Y[1] = dq_r;
        Y[2] = Math.sin(q);
        double tauHat = Y[0] * p.mHat + Y[1] * p.bHat + Y[2] * p.g0Hat;
        tauHatOut[0] = tauHat;
        return tauHat;
    }

    public static void main(String[] args) {
        Params p = new Params();
        double kP = 20.0, kD = 8.0;
        double dt = 0.001, T = 5.0;
        int steps = (int)(T / dt);

        double q = 0.0, dq = 0.0;
        double[] Y = new double[3];
        double[] tauHatOut = new double[1];
        for (int k = 0; k < steps; ++k) {
            double t = k * dt;
            double tau = controlLaw(t, q, dq, p, kP, kD, Y, tauHatOut);
            double tauHat = tauHatOut[0];

            // Simple gradient adaptation
            double error = tau - tauHat;
            double gamma = 0.1;
            p.mHat -= gamma * (-Y[0] * error);
            p.bHat -= gamma * (-Y[1] * error);
            p.g0Hat -= gamma * (-Y[2] * error);

            // Plant integration
            double ddq = (tau - p.bTrue * dq - p.g0True * Math.sin(q)) / p.mTrue;
            dq += ddq * dt;
            q += dq * dt;

            if (k % 1000 == 0) {
                System.out.printf("t=%.3f q=%.3f mHat=%.3f%n", t, q, p.mHat);
            }
        }
    }
}
